src/irr/asset/CCPUMeshPacker.cpp

```diff
@@ -44,18 +30,124 @@ std::optional<std::pair<ICPUMesh*, DrawElementsIndirectCommand_t>> CCPUMeshPacke
 		}
 	}
 
-	for (int i = 1; i < meshBuffers.size(); i++)
+	//define SVertexInputParams of packed mesh buffer
+	SVertexInputParams outVtxInputParams = meshBuffers[0]->getPipeline()->getVertexInputParams();
+	outVtxInputParams.enabledBindingFlags = 0x0001; // < binding for unified VBO
+	memset(outVtxInputParams.bindings, 0, sizeof(SVertexInputParams::bindings));
+
+	auto findAttribIndexBoundToN = [](const uint32_t N, const SVertexInputParams& attrParams) -> int
 	{
-		const auto& vxInputParamsRef = meshBuffers[0]->getPipeline()->getVertexInputParams();
-		const auto& vxInputParams = meshBuffers[i]->getPipeline()->getVertexInputParams();
+		if (N > 15)
+			return -1;
 
-		if (std::memcmp(&vxInputParamsRef, &vxInputParams, sizeof(SVertexInputParams)))
+		for (uint32_t i = 0; i < SVertexInputParams::MAX_VERTEX_ATTRIB_COUNT; i++)
 		{
-			_IRR_DEBUG_BREAK_IF(true);
-			return {};
+			const SVertexInputAttribParams& vtxInputParam = attrParams.attributes[i];
+			if (vtxInputParam.binding == N)
+				return i;
+		}
+
+		return -1;
+	};
+
+	uint16_t sharedVtxAttribsMask = outVtxInputParams.enabledAttribFlags; //< attributes that are enabled for EVERY input mesh buffers ("alive" attributes)
+	//establish which attrib bindings are shared among all mesh buffers ("alive" attributes)
+	for (size_t i = 1; i < meshBuffers.size(); i++)
+	{
+		const auto& currMBVtxInputParams = meshBuffers[i]->getPipeline()->getVertexInputParams();
+
+		if (currMBVtxInputParams.enabledAttribFlags != outVtxInputParams.enabledAttribFlags)
+		{
+			//if currently processed mesh buffer has given attrib enabled but packed mesh buffer don't, then eneble this attrib in packed mesh buffer
+			outVtxInputParams.enabledAttribFlags |= currMBVtxInputParams.enabledAttribFlags;
+
+			for (uint16_t attrBit = 0x0001, binding = 0; binding < 16; attrBit <<= 1, binding++)
+			{
+				//if currently processed mesh buffer doesn't have given attrib enabled but all of the previous mesh buffers had, then remove this attrib from `sharedVtxAttribsMask`
+				if (!(attrBit & currMBVtxInputParams.enabledAttribFlags) && (attrBit & sharedVtxAttribsMask))
+					sharedVtxAttribsMask ^= attrBit;
+
+				//set format of new attribute
+				const int attrBindingIndex = findAttribIndexBoundToN(binding, currMBVtxInputParams);
+				assert(attrBindingIndex != -1); //imposibru 
+
+				const asset::E_FORMAT formatOfNewAttr = static_cast<asset::E_FORMAT>(currMBVtxInputParams.attributes[attrBindingIndex].format);
+
+				//assign to first matching attr param
+				SVertexInputAttribParams* attrToAssign = outVtxInputParams.attributes;
+				for (uint32_t i = 0; i < SVertexInputParams::MAX_ATTR_BUF_BINDING_COUNT; i++)
+				{
+					if (attrToAssign->format == asset::EF_UNKNOWN)
+					{
+						attrToAssign->format = formatOfNewAttr;
+						break;
+					}
+
+					attrToAssign++;
+				}
+			}
```

Comment:
don't iterate over BINDINGS

BINDINGS are offsets and strides  that fetch your data

ATTRIBUTES have the formats

Reply:
it's a typo

Reply:
I actually iterate over attributes here